lib/url.js

```diff
@@ -75,355 +88,192 @@ const slashedProtocol = {
+  '#',
+  '?',
+  '/',
+  '\\'
+].map(function(v) {
+  return v.charCodeAt(0);
+}));
+// If these characters end a host name, the path will not be prepended a /.
+const _hostEndingCharactersNoPrependSlash = makeAsciiTable([
+  '<',
+  '>',
+  '"',
+  '`',
+  ' ',
+  '\r',
+  '\n',
+  '\t',
+  '{',
+  '}',
+  '|',
+  '^',
+  '`',
+  '\'',
+  '%',
+  ';'
+].map(function(v) {
+  return v.charCodeAt(0);
+}));
+
 const querystring = require('querystring');
 
+const noSerializePattern = /^(?:_protocol|_port|_href|_query|_prependSlash)$/;
+
 function urlParse(url, parseQueryString, slashesDenoteHost) {
   if (url instanceof Url) return url;
 
-  var u = new Url;
+  var u = new Url();
   u.parse(url, parseQueryString, slashesDenoteHost);
   return u;
 }
 
-Url.prototype.parse = function(url, parseQueryString, slashesDenoteHost) {
-  if (typeof url !== 'string') {
-    throw new TypeError("Parameter 'url' must be a string, not " + typeof url);
-  }
-
-  // Copy chrome, IE, opera backslash-handling behavior.
-  // Back slashes before the query string get converted to forward slashes
-  // See: https://code.google.com/p/chromium/issues/detail?id=25916
-  var queryIndex = url.indexOf('?'),
-      splitter =
-          (queryIndex !== -1 && queryIndex < url.indexOf('#')) ? '?' : '#',
-      uSplit = url.split(splitter),
-      slashRegex = /\\/g;
-  uSplit[0] = uSplit[0].replace(slashRegex, '/');
-  url = uSplit.join(splitter);
-
-  var rest = url;
-
-  // trim before proceeding.
-  // This is to support parse stuff like "  http://foo.com  \n"
-  rest = rest.trim();
-
-  if (!slashesDenoteHost && url.split('#').length === 1) {
-    // Try fast path regexp
-    var simplePath = simplePathPattern.exec(rest);
-    if (simplePath) {
-      this.path = rest;
-      this.href = rest;
-      this.pathname = simplePath[1];
-      if (simplePath[2]) {
-        this.search = simplePath[2];
-        if (parseQueryString) {
-          this.query = querystring.parse(this.search.substr(1));
-        } else {
-          this.query = this.search.substr(1);
-        }
-      } else if (parseQueryString) {
-        this.search = '';
-        this.query = {};
-      }
-      return this;
-    }
+Url.prototype.parse = function(str, parseQueryString, hostDenotesSlash) {
+  if (typeof str !== 'string') {
+    throw new TypeError(`Parameter 'url' must be a string, not ` +
+        typeof str);
   }
-
-  var proto = protocolPattern.exec(rest);
-  if (proto) {
-    proto = proto[0];
-    var lowerProto = proto.toLowerCase();
-    this.protocol = lowerProto;
-    rest = rest.substr(proto.length);
+  var start = 0;
+  var end = str.length - 1;
+
+  // Trim leading and trailing ws.
+  while (str.charCodeAt(start) <= 0x20 /*' '*/) start++;
+  var trimmedStart = start;
+  while (str.charCodeAt(end) <= 0x20 /*' '*/) end--;
+
+  start = this._parseProtocol(str, start, end);
+
+  // Javascript doesn't have host.
+  if (this._protocol !== 'javascript') {
```

Comment:
Yeah. It seems from those browsers mostly do not have a host. This may be an area where the URL Standard reflects reality whereas the RFCs do not.  E.g. https://url.spec.whatwg.org/interop/test-results/aad58a581d?select=current